awx/main/tasks/facts.py

```diff
@@ -57,6 +58,21 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
     return None
 
 
+def update_hosts(host_list):
+    if not host_list:
+        return
+    for i in range(5):
```

Comment:
Future me would love a comment or two here to understand why this retry loop is necessary.

Reply:
Sure, this is valid, and as I looked at it, I could predict other behavior problems. So I added a few unit tests so that we at least get coverage for all the code branches.